Add Sonoff SNZB-01M Smart Scene remote

Oniums opened the pull request:
## Proposed change
<!--
  Explain your proposed change below.
-->
This adds a quirk for the Sonoff SNZB-01M remote.

## Additional information
<!--
  Please include any additional information that is important to this PR.
  For example, if this PR is a potentially breaking change, mention that here.
  If this PR requires other PRs to be merged in HA Core or other projects, mention that.
  Lastly, if this PR fixes a specific issue, please include "Fixes #xxxx".
-->


## Device diagnostics
<!--
  Diagnostics data is used by ZHA unit tests to make sure quirks create the expected
  entities and we do not have regressions. For all new quirks, we require device
  diagnostics.

  You can find the diagnostics information by going to the device page, clicking the
  three dots, and then by clicking on "Download diagnostics". Drag-and-drop the
  downloaded file into this section.
-->
Without quirk:
[zha-01KAH6BZ8GBKBJXCPX8V4BB4KP-SONOFF SNZB-01M-8375561d015eac194cba643939d917ea.json](https://github.com/user-attachments/files/26387583/zha-01KAH6BZ8GBKBJXCPX8V4BB4KP-SONOFF.SNZB-01M-8375561d015eac194cba643939d917ea.json)



## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [x] Tests have been added to verify that the new code works
- [x] Device diagnostics data has been attached

Commit:
Add Sonoff SNZB-01M Smart Scene Button.

TheJulianJES reviewed (changes requested):
This PR adds a v1 quirk that doesn't use any constants in the quirk `signature` and `replacement`. This is something that all other v1 quirks do. Like, you'd wanna do `OnOff.cluster_id` instead of `0x0006`.
Some parts of the tests are also a bit unnecessary as well.

In general, we shouldn't add new v1 quirks anymore if v2 quirks also work.

I'll get back to you with more details soon, but for now, please take a look at some v2 quirks adding device automation triggers:
1. https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/sonoff/button.py
   This one doesn't use a custom cluster though, but it's a nice simple example.
2. https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/smarthjemmet/quadzigsw.py
   This one is a bit special, but does use a custom cluster, and works around a firmware issue with all the extra endpoints/clusters.
3. Maybe also see: https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/candeo/rotary_dimmer_switches.py

You'll wanna use the `QuirkBuilder`, `.replaces(SonoffButtonCluster, endpoint_id=x)` for the custom cluster, and `.device_automation_triggers(...)`.

For the test, something like https://github.com/zigpy/zha-device-handlers/blob/dev/tests/test_smarthjemmet.py would be enough. We don't need to test that the quirk definition and so on works properly, just added logic, like in the new custom cluster.

Commit:
feat(sonoff/snzb01m): migrate SNZB-01M quirk to v2 QuirkBuilder and focus tests on cluster behavior

Commit:
feat(sonoff/snzb01m): migrate SNZB-01M quirk to v2 QuirkBuilder and focus tests on cluster behavior

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

Commit:
test(sonoff): ensure v2 fixture creates endpoints 1-4 for all tests

Commit:
Merge branch 'dev' of github.com:Oniums/zha-device-handlers into dev

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

Oniums commented:
Thanks for the review and the example links — that helped a lot.
follow that i change in this PR

- Migrated the quirk to the v2 API (QuirkBuilder) instead of adding a new v1 quirk. This avoids hard-coded cluster IDs in signatures/replacements and follows the v2 pattern you pointed to.
- Implemented the device registration with QuirkBuilder and kept the custom cluster implementation (SonoffButtonCluster) intact.
- Simplified tests to only cover the added cluster logic (the attribute → event conversion), following the style used in the smarthjemmet example you pointed at. Tests now:
1. Use the zigpy_device_from_v2_quirk fixture.
2. Create endpoints 1–4 so the v2 quirk can add the custom cluster to each endpoint.
3. Avoid testing QuirkBuilder registration internals (device_automation_triggers/signature/replacement), per your guidance.

- Fixed lint/pre-commit issues reported during CI (moved imports to module top, added concise test docstrings).
- Fixed a duplicate metadata registration bug: I removed a redundant .applies_to(...) call that caused the same manufacturer/model to be registered twice (this was the root cause of the manufacturer-model uniqueness failure).
- Kept the quirk file exporting a SNZB01M variable (the QuirkBuilder result) so other code/tests can access device_automation_triggers if needed, but the tests themselves no longer assert on that registry mapping.

### Files changed
- snzb01m.py
1. QuirkBuilder registration (v2), removed redundant .applies_to().
2. SonoffButtonCluster unchanged (still handles attribute updates → listener events).
- test_sonoff.py
1. Tests reduced to cluster behavior only (attribute handling and emitted events).
2. Use zigpy_device_from_v2_quirk("SONOFF", "SNZB-01M", endpoint_ids=[1,2,3,4]).

@TheJulianJES When you have a moment,could you please review the latest changes in this PR? Thanks

diegocr commented:
I came across this PR from Google after purchasing the SNZB-01M and finding it does not work properly in ZHA ... the fourthly button does not even signal anything in the logbook, which dunno if normal or faulty device ... anyway, any ETA when this will merged/released? Thanks!

iangreenleaf commented:
> I came across this PR from Google after purchasing the SNZB-01M and finding it does not work properly in ZHA ... the fourthly button does not even signal anything in the logbook, which dunno if normal or faulty device ... anyway, any ETA when this will merged/released? Thanks!

@diegocr that is the behavior I saw as well, so it's not a faulty device. I was able to add `snzb01m.py` from this PR to my custom quirks following [this guide](https://community.home-assistant.io/t/zigbee-guide-how-to-add-setup-local-custom-zha-device-handlers-also-known-as-quirks-in-the-zha-integration/683473) and the device is working great now, so that's a :+1: from me on this PR. It'd be nice to get this merged so the device can be used without adding custom config.

diegocr commented:
Pleased to hear you have this working, will check that guide on how to install it - thank you @iangreenleaf!

TheJulianJES committed:
Merge branch 'dev'

TheJulianJES committed:
Simplify and clean up quirk and test

TheJulianJES committed:
Address review comments

TheJulianJES committed:
Don't pass `value` because it's not needed

TheJulianJES reviewed (approved):
Thanks! I've simplified the quirk and tests a bit. Everything still works fine (tested using a physical device).
I've also tested the new version from the OTA you provided and it works nicely:
- https://github.com/zigpy/zigpy-ota/issues/50

If you have any questions regarding the changes I made in the quirk, let me know.

TheJulianJES commented on tests/test_sonoff.py:
```diff
@@ -219,3 +221,54 @@ async def test_sonoff_cluster_apply_custom_configuration(zigpy_device_from_v2_qu
+
+
+@pytest.mark.parametrize("endpoint_id", [1, 2, 3, 4])
+@pytest.mark.parametrize(
+    ("value", "expected_action"),
+    [
+        (1, SHORT_PRESS),
+        (2, DOUBLE_PRESS),
+        (3, LONG_PRESS),
+        (4, TRIPLE_PRESS),
+    ],
```
This won't make a difference for the user if they use the device triggers and not directly listen to these custom `zha_event`s, but it makes sense to address. Fixed now.

TheJulianJES commented on zhaquirks/sonoff/snzb01m.py:
```diff
@@ -0,0 +1,76 @@
+    .device_automation_triggers(
+        {
+            # (SHORT_PRESS, "button_1"): {COMMAND: SHORT_PRESS, ENDPOINT_ID: 1},
+            # ...
```
I think this somewhat nicely shows what the triggers are, so I'd leave it in for now.

TheJulianJES commented on zhaquirks/sonoff/snzb01m.py:
```diff
@@ -0,0 +1,76 @@
+    .adds(SonoffButtonCluster, endpoint_id=2)
+    .adds(SonoffButtonCluster, endpoint_id=3)
+    .adds(SonoffButtonCluster, endpoint_id=4)
+    .device_automation_triggers(
+        {
+            # (SHORT_PRESS, "button_1"): {COMMAND: SHORT_PRESS, ENDPOINT_ID: 1},
+            # ...
+            (action, button): {COMMAND: action, ENDPOINT_ID: ep}
+            for ep, button in BUTTONS.items()
+            for action in (SHORT_PRESS, DOUBLE_PRESS, LONG_PRESS, TRIPLE_PRESS)
+        }
```
This won't make a difference for the user if they use the device triggers and not directly listen to these custom `zha_event`s, but it makes sense to address. Fixed now.

TheJulianJES commented on zhaquirks/sonoff/snzb01m.py:
```diff
@@ -0,0 +1,76 @@
+            manufacturer_code=None,
+        )
+
+    def _update_attribute(self, attrid, value):
+        super()._update_attribute(attrid, value)
+        if attrid == self.AttributeDefs.key_action_event.id:
+            action = ACTION_MAP.get(value)
+            if action:
+                self.listener_event(ZHA_SEND_EVENT, action, {})
```
There's no real use in including other arguments, like `value` here. The event itself is enough.

TheJulianJES commented on zhaquirks/sonoff/snzb01m.py:
```diff
@@ -0,0 +1,76 @@
+    ZHA_SEND_EVENT,
+)
+
+BUTTONS = {
+    1: BUTTON_1,
+    2: BUTTON_2,
+    3: BUTTON_3,
+    4: BUTTON_4,
+}
+
+ACTION_MAP = {
+    1: SHORT_PRESS,
+    2: DOUBLE_PRESS,
+    3: LONG_PRESS,
+    4: TRIPLE_PRESS,
+}
```
This won't make a difference for the user if they use the device triggers and not directly listen to these custom `zha_event`s, but it makes sense to address. Fixed now.

TheJulianJES committed:
Use `replaces`

TheJulianJES commented on tests/test_sonoff.py:
```diff
@@ -219,3 +221,54 @@ async def test_sonoff_cluster_apply_custom_configuration(zigpy_device_from_v2_qu
+):
+    """Correct events are emitted for each endpoint and action."""
+
+    device = zigpy_device_from_v2_quirk("SONOFF", "SNZB-01M", endpoint_ids=[1, 2, 3, 4])
+    cluster = device.endpoints[endpoint_id].sonoff_button_cluster
```
That's true, but it's good enough if we adds/replace the clusters in the quirk.

TheJulianJES commented on zhaquirks/sonoff/snzb01m.py:
```diff
@@ -0,0 +1,87 @@
+    .adds(SonoffButtonCluster, endpoint_id=1)
+    .adds(SonoffButtonCluster, endpoint_id=2)
+    .adds(SonoffButtonCluster, endpoint_id=3)
+    .adds(SonoffButtonCluster, endpoint_id=4)
```
Fixed. The cluster already seems to be part of the signature by default, so I wonder if we have an issue here where `adds` and `replaces` basically do the same...
